File: 9_排序和搜索/力扣加加/计数二分/378. 有序矩阵中第 K 小的元素.ts
/**
 * @param {number[][]} matrix
 * @param {number} k
 * @return {number}
 */
const kthSmallest = function (matrix: number[][], k: number): number {
  const count = (mid: number) => {
    let count = 0
    for (let i = 0; i < matrix.length; i++) {
      for (let j = 0; j < matrix[0].length; j++) {
        if (matrix[i][j] <= mid) count++
        else break
      }
    }
    return count
  }
  let l = matrix[0][0]
  let r = matrix[matrix.length - 1][matrix[0].length - 1] + 1

  while (l <= r) {
    const mid = ~~((l + r) / 2)
    if (count(mid) < k) l = mid + 1
    else r = mid - 1
  }

  return l
}

export default 1

console.log(
  kthSmallest(
    [
      [1, 5, 9],
      [10, 11, 13],
      [12, 13, 15],
    ],
    8
  )
)

File: 8_heap/力扣加加/三个技巧/多路归并/378. 有序矩阵中第 K 小的元素.py
from heapq import heappop, heappush, heapify
from typing import List

# 每行和每列元素均按升序排序 => 从左下角开始寻找，二分
# 每行均按升序排序，列不一定升序排序 => 每行多路归并


class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        n = len(matrix)
        pq = [(matrix[i][0], i, 0) for i in range(n)]
        heapify(pq)

        for _ in range(k - 1):
            _, x, y = heappop(pq)
            if y != n - 1:
                heappush(pq, (matrix[x][y + 1], x, y + 1))

        return heappop(pq)[0]

    def kthSmallest2(self, matrix: List[List[int]], k: int) -> int:
        n = len(matrix)

        def countNGT(mid: int) -> int:
            row, col = n - 1, 0
            res = 0
            while row >= 0 and col < n:
                if matrix[row][col] <= mid:
                    res += row + 1
                    col += 1
                else:
                    row -= 1
            return res

        left, right = matrix[0][0], matrix[-1][-1]
        while left <= right:
            mid = (left + right) // 2
            if countNGT(mid) < k:
                left = mid + 1
            else:
                right = mid - 1

        return left


print(Solution().kthSmallest2(matrix=[[1, 5, 9], [10, 11, 13], [12, 13, 15]], k=8))
